File: packages/config/src/flags.ts
export interface FeatureFlags {
    /** Enable file upload functionality */
    enableUpload: boolean;
    /** Enable vector/semantic search (requires embeddings) */
    enableSemanticSearch: boolean;
    /** Enable Firebase Analytics tracking */
    enableAnalytics: boolean;
    /** Show debug panel in the frontend UI */
    enableDebugPanel: boolean;
    /** Maximum file upload size in bytes */
    maxUploadSizeBytes: number;
}

const defaults: FeatureFlags = {
    enableUpload: true,
    enableSemanticSearch: false,
    enableAnalytics: false,
    enableDebugPanel: false,
    maxUploadSizeBytes: 10 * 1024 * 1024, // 10 MB
};

function parseBool(value: string | undefined, fallback: boolean): boolean {
    if (value === undefined || value === "") return fallback;
    return value === "true" || value === "1";
}

function parseInt(value: string | undefined, fallback: number): number {
    if (value === undefined || value === "") return fallback;
    const parsed = Number(value);
    return Number.isNaN(parsed) ? fallback : parsed;
}

/**
 * Reads feature flags from environment variables.
 * Frontend: pass `import.meta.env`
 * Backend: pass `os.environ` (via Python, not this function)
 *
 * Env var naming: `FLAG_<SCREAMING_SNAKE_CASE>`
 * e.g. `FLAG_ENABLE_UPLOAD=true`
 */
export function loadFlags(env: Record<string, string | undefined>): FeatureFlags {
    return {
        enableUpload: parseBool(env["FLAG_ENABLE_UPLOAD"], defaults.enableUpload),
        enableSemanticSearch: parseBool(env["FLAG_ENABLE_SEMANTIC_SEARCH"], defaults.enableSemanticSearch),
        enableAnalytics: parseBool(env["FLAG_ENABLE_ANALYTICS"], defaults.enableAnalytics),
        enableDebugPanel: parseBool(env["FLAG_ENABLE_DEBUG_PANEL"], defaults.enableDebugPanel),
        maxUploadSizeBytes: parseInt(env["FLAG_MAX_UPLOAD_SIZE_BYTES"], defaults.maxUploadSizeBytes),
    };
}

export { defaults as defaultFlags };

File: backend/app/flags.py
from __future__ import annotations

import os
from dataclasses import dataclass


@dataclass(frozen=True)
class FeatureFlags:
    enable_upload: bool = True
    enable_semantic_search: bool = False
    enable_analytics: bool = False
    enable_debug_panel: bool = False
    max_upload_size_bytes: int = 10 * 1024 * 1024


def _parse_bool(value: str, fallback: bool) -> bool:
    if not value:
        return fallback
    return value.lower() in ("true", "1")


def load_flags() -> FeatureFlags:
    return FeatureFlags(
        enable_upload=_parse_bool(os.environ.get("FLAG_ENABLE_UPLOAD", ""), True),
        enable_semantic_search=_parse_bool(os.environ.get("FLAG_ENABLE_SEMANTIC_SEARCH", ""), False),
        enable_analytics=_parse_bool(os.environ.get("FLAG_ENABLE_ANALYTICS", ""), False),
        enable_debug_panel=_parse_bool(os.environ.get("FLAG_ENABLE_DEBUG_PANEL", ""), False),
        max_upload_size_bytes=int(os.environ.get("FLAG_MAX_UPLOAD_SIZE_BYTES", str(10 * 1024 * 1024))),
    )


flags = load_flags()
